teps/0045-guard-finally-tasks-with-whenexpressions.md

```diff
@@ -0,0 +1,303 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Users can guard execution of `Tasks` using `WhenExpressions`, but that is not supported in `Finally Tasks`. This TEP 
+describes the need for supporting `WhenExpressions` in `Finally Tasks` not only to provide efficient guarded execution
+but also to improve the reusability of `Tasks`. Given we've recently added support for `Results` and `Status` in 
+`Finally Tasks`, this is an opportune time to enable `WhenExpressions` in `Finally Tasks`. 
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this TEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a TEP within the wider Tekton community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, users cannot guard the execution of `Finally Tasks` so they are always executed. 
+Users may want to guard the execution of `Finally Tasks` based on [`Results` from other `Tasks`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#consuming-task-execution-results-in-finally). 
+Moreover, now that [the execution status of `Tasks` is accessible in `Finally Tasks`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-execution-status-of-pipelinetask),
+they may also want to guard the execution of `Finally Tasks` based on the execution status of other `Tasks`.
+
+An example use case is a `Pipeline` author wants to send a notification, such as posting on Slack using [this catalog task](https://github.com/tektoncd/catalog/tree/master/task/send-to-channel-slack/0.1#post-a-message-to-slack), 
+when a certain `Task` in the `Pipeline` failed. To do this, one user has had to use a workaround using `Workspaces` that
+they describe [in this thread](https://tektoncd.slack.com/archives/CK3HBG7CM/p1603399989171300?thread_ts=1603376439.161500&cid=CK3HBG7CM).
+In addition, needing the workaround prevents the user from reusing the Slack catalog task as further described in [this issue](https://github.com/tektoncd/pipeline/issues/3438).
+
+We already guard `Tasks` using [`WhenExpressions`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#guard-task-execution-using-whenexpressions),
+which efficiently evaluate the criteria of executing `Tasks`. We propose supporting using `WhenExpressions` to guard 
+the execution of `Finally Tasks` as well. 
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+The primary goal of this proposal is to allow guarding execution of `Finally Tasks` using `WhenExpressions`. A secondary
+goal is to improve the [reusability](https://github.com/tektoncd/community/blob/master/design-principles.md#reusability)
```

Comment:
I would switch this and make reusability as the primary goal.

Reply:
made them "co-goals", but mentioned reusability first